File: compiler/python/state-buffer-estimate/utils/printers.py
import json

from utils.funcs     import kstr
from utils.consts    import *
from utils.datatype  import *
from layers.layer import Layer
from nets.network import Network

class Printer(object):

    #-----------------------------------------------------------------
    def __init__(self, netwk):
        self.__Network = netwk

    #-----------------------------------------------------------------
    def printNetwork(self):
        ntwk = self.__Network
        prevNl = False
        maxStateSize = 0
        layerNumMajor = 0
        layerNumMinor = 0
        self.__PrevLayer = None

        for layer in ntwk.gLayers():
            if layer.gDenseBlockStart() >= 0:
                if not prevNl:
                    print
                print (">>> Starting dense block " + str(layer.gDenseBlockStart()))
            elif layer.gTranBlockStart() >= 0:
                if not prevNl:
                    print
                print(">>> Starting tran block " + str(layer.gTranBlockStart()))

            if layer.qStoreInSB():
                inStateSize = layer.gInputStateMemWithoutBatching()
                outStateSize = layer.gOutputStateMemWithoutBatching()
                totalStateSize = inStateSize + outStateSize
                if totalStateSize > maxStateSize:
                    maxStateSize = totalStateSize
            else:
                inStateSize = layer.gInputSize()
                outStateSize = layer.gOutputSize()

            numStr = layer.gNumberStr()
            print (numStr + " " + str(layer))
            layer.m_NumStr = numStr

            prevNl = False
            if layer.gDenseBlockEnd() >= 0:
                print("<<< Ending dense block " + str(layer.gDenseBlockEnd()))
                print
                prevNl = True
            elif layer.gTranBlockEnd() >= 0:
                print("<<< Ending tran block " + str(layer.gTranBlockEnd()))
                print
                prevNl = True

            self.__PrevLayer =layer

        print("Max state size =", kstr(maxStateSize))

    #-----------------------------------------------------------------
    def printDot(self):
        ntwk = self.__Network
        f1=open(netwk.gName()+".dot", 'w')

        graphName = netwk.gName().replace("-", "_").replace(".", "_")
        print >>f1, 'digraph', graphName, "{"

        for layer in netwk.gLayers():
            print >>f1, '  ', layer.gDotIdLabel()

        print >>f1

        for layer in netwk.__Layers:
            for nextLayer in layer.gNextLayers():
                print >>f1, '  ', layer.gDotId(), '->', nextLayer.gDotId(), ';'

        print >>f1, '}'
        print >>f1


    #-----------------------------------------------------------------
    def printLevels(self):
        ntwk = self.__Network
        for level in ntwk.gLevels():
            for layer in level.gLayers():
                print(layer.gNameWithSched(),)
            print

    #-----------------------------------------------------------------
    def printSched(self):
        ntwk = self.__Network
        dataType = ntwk.gDataType()
        print(ntwk.gName(), ": data type=", dataType.gName(), " data type size=", dataType.gSizeInBytes())
        memHeader = (SCHED_MEM_FORMAT) % (
            "Layer", "Ofmap", "In", "Out",
            "Residue",
            "Batch",
            "BatchDlt",
            )
        lineFmt = ("%-70s  %s")
        fullHeader = (lineFmt) % (memHeader, "SB predecessors")
        print(fullHeader)
        hasRelu = False
        lastWasAdd = False

        for layer in ntwk.gSchedLayers():
            if layer.qReluLayer():
                hasRelu = True
            sbPreds = ""
            first=True
            for sbLayer in layer.gPrevSbLayers():
                s = sbLayer.gName()
                if not first:
                    s = "," + s
                first=False
                sbPreds += s

            if sbPreds == "":
                sbPreds = "()"
            sb = "SB" if layer.qStoreInSB() else "--"
            ss = (lineFmt) % (layer.gNameWithSchedMem(), "[" + sb + "]=" + sbPreds)

            ifaddr = layer.gIfmapAddress()
            ofaddr = layer.gOfmapAddress()
            waddr  = layer.gWeightAddress()

            if ifaddr != None or ofaddr != None or waddr != None:
                ss += " {"
                b = False
                if ifaddr != None:
                    ss += "i=" + str(ifaddr)
                    b = True
                if ofaddr != None:
                    if b: ss += ", "
                    ss += "o=" + str(ofaddr)
                    b = True
                if waddr != None:
                    if b: ss += ", "
                    ss += "w=" + str(waddr)
                    b = True
                ss += "}"

            print(ss)
            if hasRelu:
                if lastWasAdd and layer.qReluLayer():
                    print
            else:
                if layer.qAddLayer() or layer.qPoolLayer():
                    print

            lastWasAdd = layer.qAddLayer()

        print(fullHeader)

    def printJson(self, obj, filename):
        obj_json = json.dumps(obj.gJson(), sort_keys=False, indent=4, separators=(',', ': '))
        with open(filename, "w") as f:
            f.write(obj_json)

File: compiler/kcc/utils/src/printers.cpp
//import json

#include <iostream>
#include <sstream>

//from utils.funcs     import kstr
#include "consts.hpp"
#include "datatype.hpp"
#include "layer.hpp"
#include "network.hpp"
#include "layerlevel.hpp"
#include "printers.hpp"

namespace kcc {
namespace utils {

//--------------------------------------------------------
void
Printer::printNetwork()
{
    Network* ntwk = m_Network;
    bool prevNl = false;
    StateBufferAddress maxStateSize = 0;
    //kcc_int32 layerNumMajor = 0;
    //kcc_int32 layerNumMinor = 0;
    m_PrevLayer = nullptr;

    for (auto layer : ntwk->gLayers()) {
        if (layer->gDenseBlockStart() >= 0) {
            if (!prevNl) {
                std::cout << "\n";
            }
            std::cout << ">>> Starting dense block " << layer->gDenseBlockStart();
        } else if (layer->gTranBlockStart() >= 0) {
            if (!prevNl) {
                std::cout << "\n";
            }
            std::cout << ">>> Starting tran block " << layer->gTranBlockStart();
        }

        StateBufferAddress inStateSize, outStateSize, totalStateSize;

        if (layer->qStoreInSB()) {
            inStateSize = layer->gInputStateMemWithoutBatching();
            outStateSize = layer->gOutputStateMemWithoutBatching();
            totalStateSize = inStateSize + outStateSize;
            if (totalStateSize > maxStateSize) {
                maxStateSize = totalStateSize;
            }
        } else {
            inStateSize = layer->gInputSize();
            outStateSize = layer->gOutputSize();
        }

        string numStr(layer->gNumberStr());
        std::cout << numStr << " " << layer->gString();
        layer->rNumStr(numStr);

        prevNl = false;
        if (layer->gDenseBlockEnd() >= 0) {
            std::cout << "<<< Ending dense block " << layer->gDenseBlockEnd();
            std::cout << "\n";
            prevNl = true;
        } else if (layer->gTranBlockEnd() >= 0) {
            std::cout << "<<< Ending tran block " << layer->gTranBlockEnd();
            std::cout << "\n";
            prevNl = true;
        }

        m_PrevLayer = layer;
    }

    std::cout << "Max state size =" << maxStateSize;
}

//------------------------------------------------
void
Printer::printDot()
{
    Network* ntwk = m_Network;
    const string dotFileName(ntwk->gName() + ".dot");
    FILE* f1 = fopen(dotFileName.c_str(), "w");

    string graphName = ntwk->gName();
    for (auto& ch : graphName) {
        if (ch == '-' || ch == '.') {
            ch = '_';
        }
    }
    fprintf(f1, "digraph %s {\n", graphName.c_str());

    for (auto layer : ntwk->gLayers()) {
        fprintf(f1, "  %s\n", layer->gDotIdLabel().c_str());
    }

    fprintf(f1, "\n");

    for (auto layer : ntwk->gLayers()) {
        for (auto nextLayer : layer->gNextLayers()) {
            fprintf(f1, "  %s->%s;\n", layer->gDotId().c_str(), nextLayer->gDotId().c_str());
        }
    }

    fprintf(f1, "}\n\n");
}


//-----------------------------------------------------------------
void
Printer::printLevels()
{
    Network* ntwk = m_Network;
    for (auto level : ntwk->gLevels()) {
        for (auto layer : level->gLayers()) {
            std::cout << (layer->gNameWithSched()) << "\n";
        }
        std::cout << "\n";
    }
}

//-----------------------------------------------------------------
void
Printer::printSched()
{
    Network* ntwk = m_Network;
    const DataType& dataType(ntwk->gDataType());
    std::cout << ntwk->gName() << ": data type=" << dataType.gName()
              << " data type size=" << dataType.gSizeInBytes() << "\n";
    char memHeader[256];
    sprintf(memHeader, SCHED_MEM_FORMAT,
        "Layer", "Ofmap", "In", "Out",
        "Residue", "Batch", "BatchDlt");
    const char* lineFmt = "%-70s  %s";
    char fullHeader[256];
    sprintf(fullHeader, lineFmt, memHeader, "SB predecessors");
    std::cout << fullHeader;
    bool hasRelu = false;
    bool lastWasAdd = false;

    for (auto layer : ntwk->gSchedForwLayers()) {
        if (layer->qReluLayer()) {
            hasRelu = true;
        }
        string sbPreds = "";
        bool first=true;
        for (auto sbLayer : layer->gPrevSbLayers()) {
            string s = sbLayer->gName();
            if (! first) {
                s = "," + s;
            }
            first=false;
            sbPreds += s;
        }

        if (sbPreds == "") {
            sbPreds = "()";
        }
        const char* sb = layer->qStoreInSB() ? "SB" : "--";
        char line[256];
        sprintf(line, lineFmt, layer->gNameWithSchedMem().c_str(), sb, sbPreds.c_str());
        std::stringstream ss;
        ss << line;

        StateBufferAddress ifaddr = layer->gIfmapAddress();
        StateBufferAddress ofaddr = layer->gOfmapAddress();
        StateBufferAddress waddr  = layer->gWeightAddress();

        if (ifaddr != StateBufferAddress_Invalid || ofaddr != StateBufferAddress_Invalid ||
           waddr != StateBufferAddress_Invalid)
        {
            ss << " {";
            bool b = false;
            if (ifaddr != StateBufferAddress_Invalid) {
                ss << "i=" << ifaddr;
                b = true;
            }
            if (ofaddr != StateBufferAddress_Invalid) {
                if (b) {
                    ss << ", ";
                }
                ss << "o=" << ofaddr;
                b = true;
            }
            if (waddr != StateBufferAddress_Invalid) {
                if (b) {
                    ss << ", ";
                }
                ss << "w=" << waddr;
                b = true;
            }
            ss << "}";
        }

        std::cout << ss.str() << "\n";
        if (hasRelu) {
            if (lastWasAdd && layer->qReluLayer()) {
                std::cout << "\n";
            }
        } else {
            if (layer->qAddLayer() || layer->qPoolLayer()) {
                std::cout << "\n";
            }
        }

        lastWasAdd = layer->qAddLayer();

    }
    std::cout << fullHeader << "\n";
}


}}


